### ISSUE TITLE
'Search' object has no attribute 'model'

### ISSUE THREAD

**Issue by DusanMadar** [barseghyanartur/django-elasticsearch-dsl-drf#62]
I am aware of https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/issues/46, but downgrading `elasticsearch` didn't help.

I get this error following http://django-elasticsearch-dsl-drf.readthedocs.io/en/0.10/quick_start.html.

```
ERROR 2018-07-11 12:11:08,036 exception Internal Server Error: /api/v2/search/
Traceback (most recent call last):
  File "/home/my_username/.virtualenvs/my_venv_name/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/home/my_username/.virtualenvs/my_venv_name/lib/python3.6/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/my_username/.virtualenvs/my_venv_name/lib/python3.6/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/my_username/.virtualenvs/my_venv_name/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/my_username/.virtualenvs/my_venv_name/lib/python3.6/site-packages/rest_framework/viewsets.py", line 90, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/my_username/.virtualenvs/my_venv_name/lib/python3.6/site-packages/rest_framework/views.py", line 489, in dispatch
    response = self.handle_exception(exc)
  File "/home/my_username/.virtualenvs/my_venv_name/lib/python3.6/site-packages/rest_framework/views.py", line 449, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/home/my_username/.virtualenvs/my_venv_name/lib/python3.6/site-packages/rest_framework/views.py", line 477, in dispatch
    self.initial(request, *args, **kwargs)
  File "/home/my_username/.virtualenvs/my_venv_name/lib/python3.6/site-packages/rest_framework/views.py", line 395, in initial
    self.check_permissions(request)
  File "/home/my_username/.virtualenvs/my_venv_name/lib/python3.6/site-packages/rest_framework/views.py", line 328, in check_permissions
    if not permission.has_permission(request, self):
  File "/home/my_username/.virtualenvs/my_venv_name/lib/python3.6/site-packages/rest_framework/permissions.py", line 133, in has_permission
    perms = self.get_required_permissions(request.method, queryset.model)
AttributeError: 'Search' object has no attribute 'model'
```

Packages I have installed

```
Django==1.11.10
django-elasticsearch-dsl==0.5.0
django-elasticsearch-dsl-drf==0.10
djangorestframework==3.6.4
elasticsearch==6.3.0
elasticsearch-dsl==6.1.0
```

**Comment by barseghyanartur** [comment]
When did you get this error? What did you do?

**Comment by DusanMadar** [comment]
I get this error following http://django-elasticsearch-dsl-drf.readthedocs.io/en/0.10/quick_start.html.

**Comment by DusanMadar** [comment]
Overriding `get_queryset()` does the trick:

```
from django_elasticsearch_dsl_drf.filter_backends import SearchFilterBackend
from django_elasticsearch_dsl_drf.views import BaseDocumentViewSet

from my_app.documents import MyDocument
from my_app.serializers import MyDocumentSearchSerializer


class MyDocumentSearchViewSet(BaseDocumentViewSet):
    document = MyDocument
    serializer_class = MyDocumentSearchSerializer

    filter_backends = [
        SearchFilterBackend,
    ]

    search_fields = (
        'name',
    )

    def get_queryset(self):
        qs = super().get_queryset()
        qs.model = self.document._doc_type.model
        return qs
```

**Comment by barseghyanartur** [comment]
Have you declated `MyModel` in the Meta of `MyDocument`?
Have you declated `MyDocument` in the Meta of `MySerializer`?

**Comment by DusanMadar** [comment]
Yes for both.

**Comment by barseghyanartur** [comment]
It [should not be necessary](https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/blob/master/examples/simple/search_indexes/viewsets/author.py) to override `get_queryset`.

Could you run the example project to see if it works for you?

**Comment by barseghyanartur** [comment]
@DusanMadar:

Did you check it with example project?

**Comment by DusanMadar** [comment]
Not yet. I will let you know when I get to it.

**Comment by DusanMadar** [comment]
OK, the example project works. It uses `elasticsearch-dsl==5.3.0`. With the 'default' dependencies I am able to load http://localhost:8001/search/authors/ and see data. Though, when I install `elasticsearch-dsl==6.1.0`  http://localhost:8001/search/authors/ returns 

```
{
  count: 0,
  next: null,
  previous: null,
  facets: { },
  results: [ ]
}
```

**Comment by barseghyanartur** [comment]
Did you reindex after switching to 6.x? Indexes are often not compatible between version.

**Comment by DusanMadar** [comment]
Right, after double checking everything with docs and the example app for like a million times, the whole `'Search' object has no attribute 'model'` boils down to `djangorestframework-jwt==1.11.0` (https://github.com/GetBlimp/django-rest-framework-jwt) which I use as `DEFAULT_PERMISSION_CLASSES` in `REST_FRAMEWORK` config to be able to authenticate API calls with JWT.

```
REST_FRAMEWORK = {
    'DEFAULT_PERMISSION_CLASSES': [
        'rest_framework.permissions.DjangoModelPermissionsOrAnonReadOnly',
    ],
   ...
```

**Comment by Fettah** [comment]
that look like a permission issue: 

I hope adding this to your view will do the trick.
`from rest_framework import permissions
from rest_framework.decorators import api_view, permission_classes


@permission_classes((permissions.AllowAny,))
class MyAnnoyingStupidView(DocumentViewSet):
    pass # maybe some weird code here as well :D
`

**Comment by barseghyanartur** [comment]
@DusanMadar, @Fettah: 

Thanks for research! I'll check it asap.

**Comment by barseghyanartur** [comment]
@DusanMadar, @Fettah:

The ``get_queryset`` method of the ``BaseDocumentViewSet`` has been modified in the following way.

```python

    def get_queryset(self):
        """Get queryset."""
        queryset = self.search.query()
        # Model- and object-permissions of the Django REST framework (
        # at the moment of writing they are ``DjangoModelPermissions``,
        # ``DjangoModelPermissionsOrAnonReadOnly`` and
        # ``DjangoObjectPermissions``) require ``model`` attribute to be
        # present in the queryset. Unfortunately we don't have that here.
        # The following approach seems to fix that (pretty well), since
        # model and object permissions would work out of the box (for the
        # correspondent Django model/object). Alternative ways to solve this
        # issue are: (a) set the ``_ignore_model_permissions`` to True on the
        # ``BaseDocumentViewSet`` or (b) provide alternative permission classes
        # that are almost identical to the above mentioned classes with
        # the only difference that they know how to extract the model from the
        # given queryset. If you think that chosen solution is incorrect,
        # please make an issue or submit a pull request explaining the
        # disadvantages (and ideally - propose  a better solution). Couple of
        # pros for current solution: (1) works out of the box, (2) does not
        # require modifications of current permissions (which would mean we
        # would have to keep up with permission changes of the DRF).
        queryset.model = self.document._doc_type.model
        return queryset
```

It seems to fix the issue for me. Any remarks/objections/concerns are welcome.

**Comment by DusanMadar** [comment]
@barseghyanartur Go for it. And good comment!

**Comment by barseghyanartur** [comment]
Solved in 0.12.
